### ISSUE TITLE
ufunc 'create_collection' not supported for the input types

### ISSUE THREAD

**Issue by mwtoews** [shapely/shapely#2098]
This is a recent issue identified using nightly numpy binary wheels with the main branch of shapely.

The error looks like this:
```
=========================== test session starts ============================
platform linux -- Python 3.10.12, pytest-8.2.2, pluggy-1.5.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /home/mtoews/src/shapely/shapely
configfile: setup.cfg
plugins: syrupy-4.6.1, anyio-4.4.0, dotenv-0.5.2, flaky-3.8.1, hypothesis-6.108.5, cov-5.0.0, xdist-3.6.1, benchmark-4.0.0
collected 9 items / 1 error                                                

================================== ERRORS ==================================
__________ ERROR collecting shapely/tests/geometry/test_coords.py __________
shapely/tests/geometry/test_coords.py:5: in <module>
    from shapely.tests.common import line_string, line_string_z, point, point_z
shapely/tests/common.py:26: in <module>
    multi_point = shapely.MultiPoint([(0, 0), (1, 2)])
shapely/geometry/multipoint.py:62: in __new__
    return shapely.multipoints(subs)
shapely/decorators.py:77: in wrapped
    return func(*args, **kwargs)
shapely/creation.py:433: in multipoints
    return lib.create_collection(geometries, typ, out=out, **kwargs)
E   TypeError: ufunc 'create_collection' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''
========================= short test summary info ==========================
ERROR shapely/tests/geometry/test_coords.py - TypeError: ufunc 'create_collection' not supported for the input types,...
!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!
============================= 1 error in 0.41s =============================
```
I've bisected the issue to https://github.com/numpy/numpy/commit/f3b669d391cfe660b064af052973b9bff9372dd8 from https://github.com/numpy/numpy/pull/26905 (ping @seberg) which is related to [NEP 50](https://numpy.org/neps/nep-0050-scalar-promotion.html).

**Comment by mwtoews** [comment]
Here is a short reproducer:
```pycon
>>> import shapely
>>> from shapely import GeometryType
>>> pts = shapely.from_wkt(["POINT(1 2)", "POINT(3 4)"])
>>> shapely.lib.create_collection(pts, GeometryType.MULTIPOINT)
```
where the last line would normally return `<MULTIPOINT ((1 2), (3 4))>` before the referenced numpy commit or the error:
> TypeError: ufunc 'create_collection' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

**Comment by seberg** [comment]
Makes sense, I guess intenums may be int subclasses, is it a problem to cast to `int()` to make it `i` compatible (rather than `int64`/`intp`?)

**Comment by mwtoews** [comment]
Thanks @seberg for taking a look; a few follow-up items/questions:

- Firstly, yes casting `int(GeometryType.MULTIPOINT)` before sending to the ufunc is a possible change we can make.
- But why does (e.g.) `np.sqrt(GeometryType.MULTIPOINT)` cast the enum to float64 without error? Is there a change that we could  make on the C-side?
- For what it's worth, I see `np.result_type(GeometryType.MULTIPOINT)` return `dtype('int64')`.
- Should there have been a warning that this API change was about to happen? I didn't see any before the breaking commit.

**Comment by seberg** [comment]
You can support the dfault int in C mainly.  It's a pretty niche change and a warning is hard, because it is rare that it will trigger any change.  Since most ufuncs will support the default integer explicitly.

**Comment by jorisvandenbossche** [comment]
@mwtoews thanks for the investigation and @seberg for the explanation. 

I was quickly checking in which places we should need such a cast, and opened a PR for that -> https://github.com/shapely/shapely/pull/2099

We do have a set of other enums for modelling parameter options, and in all those cases it seems we were already casting the enum to an int. For example:

https://github.com/shapely/shapely/blob/9a8bf73ecf2ef51d684b5866ab5718220b402640/shapely/_geometry.py#L842-L852

or the various buffer arguments:

https://github.com/shapely/shapely/blob/9a8bf73ecf2ef51d684b5866ab5718220b402640/shapely/constructive.py#L181-L190

(although I see we consistently use `np.intc(..)` for the casting while in the PR I just opened I used a python integer `int(..)`, will change that for consistency)

**Comment by seberg** [comment]
Sounds right.  Is this a big enough problem to worry about delaying things in NumPy (also since 2.1 is probably be going to be relatively soon for Python 3.13 support).

**Comment by jorisvandenbossche** [comment]
I don't think so, we can always include this small fix in another 2.0.x release

**Comment by AlessandroMondin** [comment]
Same problem with numpy 2.1.0 (released yesterday)

**Comment by ethan-moss** [comment]
@AlessandroMondin - as a short-term solution, reverting back to numpy 2.0.1 has at least patched the issue for me. Hope that helps!

**Comment by jorisvandenbossche** [comment]
Shapely 2.0.6 is released that includes the fix (it's on pypi, conda-forge can still take a few hours)

**Comment by ChrisBarker-NOAA** [comment]
I jsut found this issue with:

```
python 3.11
shapely 2.0.5
numpy 2.1.0
```

all conda-forge builds.

updating to shapely 2.0.6 fixed it -- nice work!

But maybe the 2.0.5 builds on conda-forge should be marked as broken?

https://github.com/conda-forge/admin-requests?tab=readme-ov-file#mark-packages-as-broken-on-conda-forge

**Comment by martinfleis** [comment]
They're not broken, just not compatible with the latest shapely. Following this logic, we would have to mark as broken all releases prior to 2.0.6.

**Comment by ChrisBarker-NOAA** [comment]
Got it -- but the dependencies are wrong, yes? 

That can be updated with:

https://prefix.dev/blog/repodata_patching

https://github.com/conda-forge/conda-forge-repodata-patches-feedstock

I think shapely 2.0.5 should have:

    `numpy >=1.19,<2.1`

rather than:

    `numpy >=1.19,<3`

I could help with this if need be, though I don't have permissions on the feedstock, so maybe not?

**Comment by jorisvandenbossche** [comment]
Yeah, on conda-forge the numpy 2.0 support was added later, and for packages built before that, conda-forge automatically ads a numpy<2 constraint (even though shapely itself didn't have that). 
So in practice, it seems that indeed only shapely 2.0.5 could be patched with a numpy<2.1 (shapely 2.0.4 still has a numpy<2 upper pin). I suppose opening a PR for that in https://github.com/conda-forge/conda-forge-repodata-patches-feedstock is welcome (I don't have any permissions their either, it will need to be reviewed by a core conda-forge team member anyway)
